Implement ArgumentCoder for SharedMemoryHandle

donny-dont opened the pull request:
#### 4efa5e604a3679ed00add70c7451c7dfb50b0a59
<pre>
Implement ArgumentCoder for SharedMemoryHandle
<a href="https://bugs.webkit.org/show_bug.cgi?id=256402">https://bugs.webkit.org/show_bug.cgi?id=256402</a>

Reviewed by Darin Adler.

Move serialization of `SharedMemoryHandle` to use an `ArgumentCoder`.
This patch prepares `SharedMemoryHandle` for being generated through
a `serialization.in` file. The generator cannot currently handle move
semantics so it was not able to be used yet.

* Source/WebKit/Platform/SharedMemory.cpp:
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/Shared/ShareableBitmap.serialization.in:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:

Canonical link: <a href="https://commits.webkit.org/263781@main">https://commits.webkit.org/263781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf8d951c2bac06210b051cfb247ca8b006083f7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5738 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5901 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7295 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6130 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5869 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7901 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7351 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3366 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12939 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5234 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5242 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7298 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5687 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4659 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5133 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1351 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9249 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5493 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -26,6 +26,7 @@
 
 #pragma once
 
+#include <wtf/ArgumentCoder.h>
```
Let’s use a forward declaration of the `ArgumentCoder` class template instead of an include, to keep build times in check.

donny-dont replied:
Fixed

darinadler commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -151,4 +119,50 @@ class SharedMemory : public ThreadSafeRefCounted<SharedMemory> {
 #endif
 };
 
+class SharedMemoryHandle {
```
What’s the rationale for moving the `SharedMemoryHandle` class down within the header? It makes this patch harder to review, so maybe leave it out?

donny-dont replied:
Will fix

donny-dont replied:
Moved back to its original position

darinadler commented on Source/WebKit/Platform/SharedMemory.cpp:
```diff
@@ -64,13 +65,46 @@ Ref<SharedBuffer> SharedMemory::createSharedBuffer(size_t dataSize) const
+
+void ArgumentCoder<WebKit::SharedMemoryHandle>::encode(Encoder& encoder, const WebKit::SharedMemoryHandle& handle)
+{
+    encoder << WTFMove(handle.m_handle);
```
This seems wrong. If the handle was passed with `const&` then it’s wrong to use `WTFMove` on one of its data members.

donny-dont replied:
The handle is mutable so it’s valid. Later on I want to remove that.

donny-dont commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -69,25 +80,19 @@ class SharedMemoryHandle {
 #endif
 
 private:
-#if USE(UNIX_DOMAIN_SOCKETS)
-    mutable UnixFileDescriptor m_handle;
-#elif OS(DARWIN)
-    mutable MachSendRight m_handle;
-#elif OS(WINDOWS)
-    mutable Win32Handle m_handle;
-#endif
```
@darinadler the `m_handle` is marked as `mutable` so the `WTFMove` is valid in the encoder. At this time I'm not able to get rid of the `mutable` but plan to continue on and try to remove it in the future.

darinadler commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -151,4 +156,14 @@ class SharedMemory : public ThreadSafeRefCounted<SharedMemory> {
+namespace IPC {
+
+template<> struct ArgumentCoder<WebKit::SharedMemoryHandle, void> {
+    static void encode(Encoder&, const WebKit::SharedMemoryHandle&);
```
Even before we change the `mutable` aspect of `SharedMemoryHandle`, I suggest we follow up by removing this function and making changes everywhere to get things working. Maybe that’s exactly what you mean by getting rid of the need for `mutable`. To state the obvious, the real problem here is not the use of `mutable`, but rather than lack of clarity about what happens to these handles during the encoding process.

donny-dont committed:
Implement ArgumentCoder for SharedMemoryHandle
https://bugs.webkit.org/show_bug.cgi?id=256402

Reviewed by Darin Adler.

Move serialization of `SharedMemoryHandle` to use an `ArgumentCoder`.
This patch prepares `SharedMemoryHandle` for being generated through
a `serialization.in` file. The generator cannot currently handle move
semantics so it was not able to be used yet.

* Source/WebKit/Platform/SharedMemory.cpp:
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/Shared/ShareableBitmap.serialization.in:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:

Canonical link: https://commits.webkit.org/263781@main
